src/sema/unused_variable.rs

```diff
@@ -0,0 +1,409 @@
+
+        Expression::ExternalFunction {
+            loc: _,
+            ty: _,
+            address,
+            function_no,
+        } => {
+            used_variable(ns, address, symtable);
+            if ns.functions[*function_no].is_accessor {
+                let body = ns.functions[*function_no].body[0].clone();
+                if let Statement::Return(_, exprs) = &body {
+                    used_variable(ns, &exprs[0], symtable);
+                }
+            }
+        }
+
+        Expression::Builtin(_, _, expr_type, args) => match expr_type {
+            Builtin::ArrayPush => {
+                assigned_variable(ns, &args[0], symtable);
+                if args.len() > 1 {
+                    used_variable(ns, &args[1], symtable);
+                }
+            }
+
+            Builtin::ArrayPop => {
+                used_variable(ns, &args[0], symtable);
+            }
+            _ => {}
+        },
+
+        Expression::DynamicArrayPush(_, array, _, arg) => {
+            assigned_variable(ns, array, symtable);
+            used_variable(ns, arg, symtable);
+        }
+
+        Expression::DynamicArrayPop(_, array, _) => {
+            used_variable(ns, array, symtable);
+        }
+
+        _ => {}
+    }
+}
+
+/// Marks as used variables that appear in an expression with right and left hand side.
+pub fn check_var_usage_expression(
+    ns: &mut Namespace,
+    left: &Expression,
+    right: &Expression,
+    symtable: &mut Symtable,
+) {
+    used_variable(ns, left, symtable);
+    used_variable(ns, right, symtable);
+}
+
+/// Generate warnings for unused varibles
+fn generate_unused_warning(loc: Loc, text: &str, notes: Vec<Note>) -> Diagnostic {
+    Diagnostic {
+        level: Level::Warning,
+        ty: ErrorType::Warning,
+        pos: Some(loc),
+        message: text.parse().unwrap(),
+        notes,
+    }
+}
+
+/// Emit different warning types according to the state variable usage
+pub fn emit_warning_local_variable(variable: &symtable::Variable) -> Option<Diagnostic> {
+    match &variable.usage_type {
+        VariableUsage::Parameter => {
+            if !variable.read {
+                return Some(generate_unused_warning(
+                    variable.id.loc,
+                    &format!(
+                        "Function parameter '{}' has never been read.",
+                        variable.id.name
+                    ),
+                    vec![],
+                ));
+            }
+            None
+        }
+
+        VariableUsage::ReturnVariable => {
+            if !variable.assigned {
+                return Some(generate_unused_warning(
+                    variable.id.loc,
+                    &format!(
+                        "Return variable '{}' has never been assigned",
+                        variable.id.name
+                    ),
+                    vec![],
+                ));
+            }
+            None
+        }
+
+        VariableUsage::StateVariable => {
```

Comment:
State variables are not local variables. The name is misleading. Please, fix this @LucasSte.

Reply:
Fixed